File: app/src/main/java/org/techtown/samplerecorder/AudioTrack.java
package org.techtown.samplerecorder;

import android.media.AudioAttributes;
import android.media.AudioFormat;

import java.nio.ShortBuffer;

public class AudioTrack {

    public static int dataMax;

    private android.media.AudioTrack audioTrack = null;
    private Thread playThread = null;
    private ShortBuffer shortBuffer = null;
    private short[] audioData = null;
    private int capacity_buffer, len_audioData;

    public void init(int sampleRate, int bufferSize) {
//        myLog.d("method activate");

        capacity_buffer = sampleRate * 240;  // stored buffer size (60s)
        shortBuffer = ShortBuffer.allocate(capacity_buffer);

        audioData = new short[bufferSize];
        len_audioData = audioData.length;
    }

    public void play(int type, int channel, int sampleRate, Queue queue) {
//        myLog.d("method activate");

        if (audioTrack == null) {
            audioTrack = new android.media.AudioTrack.Builder()
                    .setAudioAttributes(new AudioAttributes.Builder()
                            .setUsage(AudioAttributes.USAGE_MEDIA)
                            .setContentType(type)
                            .build())
                    .setAudioFormat(new AudioFormat.Builder()
                            .setEncoding(AudioFormat.ENCODING_PCM_16BIT)
                            .setSampleRate(sampleRate)
                            .setChannelMask(channel)
                            .build())
                    .setBufferSizeInBytes(audioData.length)
                    .build();
        }

        playThread = new Thread(new Runnable() {
            @Override
            public void run() {

                audioTrack.play();

                shortBuffer = queue.dequeue();
                shortBuffer.position(0);

                while (MainActivity.isPlaying) {
                    shortBuffer.get(audioData, 0, len_audioData);  // shortBuffer -> audioData
                    audioTrack.write(audioData, 0, len_audioData);  // audioData -> audioTrack

                    dataMax = 0;
                    for (int i = 0; i < audioData.length; i++) {
                        if (Math.abs(audioData[i]) >= dataMax) {
                            dataMax = Math.abs(audioData[i]);
                        }
                    }

                    if ((audioData[audioData.length - 1] == AudioRecord.lastData_1) && (audioData[audioData.length - 2] == AudioRecord.lastData_2)) {
                        MainActivity.autoStop = true;
                        break;
                    }

                }
            }

        });
        playThread.start();
    }

    public void stop() {
//        myLog.d("method activate");

        if (audioTrack != null && audioTrack.getState() != android.media.AudioTrack.STATE_UNINITIALIZED) {
            if (audioTrack.getPlayState() != android.media.AudioTrack.PLAYSTATE_STOPPED) {
                try {
                    audioTrack.stop();

                } catch (IllegalStateException e) {
                    e.printStackTrace();
                }

                audioTrack.release();
                audioTrack = null;
                playThread = null;
            }
        }
    }

    public void release() {
//        myLog.d("method activate");
    }
}

File: app/src/main/java/org/techtown/samplerecorder/Main/AudioTrack.kt
package org.techtown.samplerecorder.Main

import android.media.AudioAttributes
import android.media.AudioFormat
import android.media.AudioTrack
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import org.techtown.samplerecorder.AppModule.dataToShort
import org.techtown.samplerecorder.MainActivity.Companion.autoStop
import org.techtown.samplerecorder.MainActivity.Companion.bufferSize
import org.techtown.samplerecorder.MainActivity.Companion.isPlaying
import org.techtown.samplerecorder.MainActivity.Companion.playChannel
import org.techtown.samplerecorder.MainActivity.Companion.playRate
import org.techtown.samplerecorder.MainActivity.Companion.type

class AudioTrack {
    private val TAG = this.javaClass.simpleName

    private var audioTrack: AudioTrack? = null
    private var job: Job? = null

    fun play(queue: Queue) {
        if (audioTrack == null) {
            audioTrack = AudioTrack.Builder()
                .setAudioAttributes(
                    AudioAttributes.Builder()
                        .setUsage(type)
                        .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
                        .build()
                )
                .setAudioFormat(
                    AudioFormat.Builder()
                        .setEncoding(AudioFormat.ENCODING_PCM_16BIT)
                        .setSampleRate(playRate)
                        .setChannelMask(playChannel)
                        .build()
                )
                .setBufferSizeInBytes(bufferSize * 2)
                .build()
        }
        var audioData: ByteArray?

        job = CoroutineScope(Dispatchers.IO).launch {
            audioTrack!!.play()
            while (isPlaying) {
                audioData = queue.dequeue()
                audioTrack!!.write(audioData!!, 0, bufferSize)

                if (queue.isEmpty) {
                    autoStop = true
                    queue.copy()
                    break
                }

                playWave = 0  // Waveform
                for (i in audioData!!.indices) playWave = dataToShort(audioData)
            }
        }
    }

    fun stop() {
        if (audioTrack != null) {
            audioTrack!!.stop()
            audioTrack!!.release()
            audioTrack = null
            job = null
        }
    }

    companion object {
        var playWave = 0
    }
}